### PR TITLE
[Remote code] Add functionality to run remote models, schedulers, pipelines

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds code that allows to run remote models, schedulers, and pipelines.
You can try it out by looking at the tests and how these two example repos are structured:
- https://huggingface.co/hf-internal-testing/tiny-sdxl-custom-all
- https://huggingface.co/hf-internal-testing/tiny-sdxl-custom-components

### CODE REVIEW

File: src/diffusers/configuration_utils.py [hunk 1]
```diff
@@ -485,10 +485,18 @@ def extract_init_dict(cls, config_dict, **kwargs):
...
+            and orig_cls_name != cls.__name__
+            and hasattr(diffusers_library, orig_cls_name)
+        ):
             orig_cls = getattr(diffusers_library, orig_cls_name)
             unexpected_keys_from_orig = cls._get_init_keys(orig_cls) - expected_keys
             config_dict = {k: v for k, v in config_dict.items() if k not in unexpected_keys_from_orig}
+        elif not isinstance(orig_cls_name, str) and not isinstance(orig_cls_name, (list, tuple)):
+            raise ValueError(
+                "Make sure that the `_class_name` is of type string or list of string (for custom pipelines)."
+            )
```
Comment (sayakpaul) [inline src/diffusers/configuration_utils.py:499]: List of strings to support the multiple components of a pipeline?

File: src/diffusers/pipelines/pipeline_utils.py [hunk 2]
```diff
@@ -323,19 +332,35 @@ def get_class_obj_and_candidates(library_name, class_name, importable_classes, p
...
-    class_obj, config, load_connected_pipeline=False, custom_pipeline=None, cache_dir=None, revision=None
+    class_obj,
+    config,
+    load_connected_pipeline=False,
+    custom_pipeline=None,
+    hub_repo_id=None,
+    hub_revision=None,
+    class_name=None,
+    cache_dir=None,
+    revision=None,
```
Comment (sayakpaul) [inline src/diffusers/pipelines/pipeline_utils.py:343]: I suggest we keep these as `**kwargs`. Instead of `hub_repo_id` or `hub_revision`, we could also opt for `repo_id` or `revision`, respectively following what we do across the library for Hub related utilities.

File: tests/pipelines/test_pipelines.py [hunk 3]
```diff
@@ -862,6 +862,58 @@ def test_run_custom_pipeline(self):
...
+        assert pipeline.__class__.__name__ == "StableDiffusionXLPipeline"
+
+        pipeline = pipeline.to(torch_device)
+        images = pipeline("test", num_inference_steps=2, output_type="np")[0]
+
+        assert images.shape == (1, 64, 64, 3)
+
+        # Check that only loading custom componets "my_unet", "my_scheduler" and explicit custom pipeline works
+        pipeline = DiffusionPipeline.from_pretrained(
+            "/home/patrick/tiny-stable-diffusion-xl-pipe", custom_pipeline="my_pipeline"
```
Comment (sayakpaul) [inline tests/pipelines/test_pipelines.py:886]: Ckpt path needs to be changed.

File: tests/pipelines/test_pipelines.py [hunk 4]
```diff
@@ -862,6 +862,58 @@ def test_run_custom_pipeline(self):
...
+        assert pipeline.config.unet == ("diffusers_modules.local.my_unet_model", "MyUNetModel")
+        assert pipeline.config.scheduler == ("diffusers_modules.local.my_scheduler", "MyScheduler")
+        assert pipeline.__class__.__name__ == "MyPipeline"
+
+        pipeline = pipeline.to(torch_device)
+        images = pipeline("test", num_inference_steps=2, output_type="np")[0]
+
+        assert images.shape == (1, 64, 64, 3)
+
+    def test_remote_auto_custom_pipe(self):
```
Comment (sayakpaul) [inline tests/pipelines/test_pipelines.py:898]: If I understand correctly, this test checks if a pipeline repo containing all custom components can be successfully loaded. If so, let's change to a better test name: `test_pipeline_with_all_custom_components()`.

File: tests/pipelines/test_pipelines.py [hunk 5]
```diff
@@ -862,6 +862,58 @@ def test_run_custom_pipeline(self):
...
+        # make sure that trust remote code has to be passed
+        with self.assertRaises(ValueError):
+            pipeline = DiffusionPipeline.from_pretrained("hf-internal-testing/tiny-sdxl-custom-all")
+
+        # Check that only loading custom componets "my_unet", "my_scheduler" and auto custom pipeline works
```
Comment (sayakpaul) [inline tests/pipelines/test_pipelines.py:903]: Not sure where the "auto" component is being leveraged here.

File: tests/pipelines/test_pipelines.py [hunk 6]
```diff
@@ -862,6 +862,58 @@ def test_run_custom_pipeline(self):
...
+            pipeline = DiffusionPipeline.from_pretrained("hf-internal-testing/tiny-sdxl-custom-components")
+
+        # Check that only loading custom componets "my_unet", "my_scheduler" works
+        pipeline = DiffusionPipeline.from_pretrained(
+            "hf-internal-testing/tiny-sdxl-custom-components", trust_remote_code=True
+        )
+
+        assert pipeline.config.unet == ("diffusers_modules.local.my_unet_model", "MyUNetModel")
+        assert pipeline.config.scheduler == ("diffusers_modules.local.my_scheduler", "MyScheduler")
+        assert pipeline.__class__.__name__ == "StableDiffusionXLPipeline"
```
Comment (sayakpaul) [inline tests/pipelines/test_pipelines.py:877]: Why should we expect this to be the class name? 

As per [this line](https://huggingface.co/hf-internal-testing/tiny-sdxl-custom-components/blob/main/my_pipeline.py#L91), this should be `"MyPipeline"` no?

File: tests/pipelines/test_pipelines.py (hunk 5 above)
Comment (sayakpaul) [reply tests/pipelines/test_pipelines.py:903]: Ah got it. So, when we don't specify `custom_pipeline`, we automatically pick the available pipeline module from the pipeline repo should it be available.

**Review by sayakpaul — APPROVED** [review]
Wow, that was pretty clean! 

Can we also add a nice doc as I believe this will be a very powerful feature. We could readily test it out with Show-1: https://github.com/showlab/Show-1. 

Also, can we add a test to see if a custom pipeline (loaded with `trust_remote_code=True` can work seamlessly with a legacy component from the library? 

For example:

```python
from diffusers import DiffusionPipeline
from diffusers import UniPCMultistepScheduler

pipeline = DiffusionPipeline.from_pretrained("hf-internal-testing/tiny-sdxl-custom-components", trust_remote_code=True)

pipeline.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config)

pipeline = pipeline.to(torch_device)
images = pipeline("test", num_inference_steps=2, output_type="np")[0]

```

**Comment by sayakpaul** [comment]
Hmm, I guess it's okay for the first shipment but it won't support loading of a custom UNet in isolation I think. 

What if I only have a model repo on the Hub similar to the ones you have for Transformers. I guess for that we need Autoclasses.

**Comment by sayakpaul** [comment]
Have been playing with this PR to support a bit more complicated pipelines such as: https://huggingface.co/showlab/show-1-base. Took a while to understand how the pipeline repository should be structured. So, documenting everything here. 

If your pipeline has custom components that `diffusers` don't support already, you need to accompany the Python modules that implement them. These custom components could be VAE, UNet, scheduler, etc. For the text encoder, we rely on `transformers` anyway. So, that should be handled separately (more info [here](https://huggingface.co/docs/transformers/custom_models)). The pipeline code itself can be custom as well. 

In case of `"showlab/show-1-base"`, we have a custom UNet and a custom pipeline (`TextToVideoIFPipeline`). For convenience, let's call the UNet `ShowOneUNet3DConditionModel`. 

`"showlab/show-1-base"` already provides the checkpoints in the `diffusers` format, which is great. So, let's start loading up the components which are already well-supported:

1. Text encoder: 

```python
from transformers import T5Tokenizer, T5EncoderModel

pipe_id = "showlab/show-1-base"
tokenizer = T5Tokenizer.from_pretrained(pipe_id, subfolder="tokenizer")
text_encoder = T5EncoderModel.from_pretrained(pipe_id, subfolder="text_encoder")
```

2. Scheduler:

```python
from diffusers import DPMSolverMultistepScheduler

scheduler = DPMSolverMultistepScheduler.from_pretrained(pipe_id, subfolder="scheduler")
```

3. Image feature extractor:

```python
from transformers import CLIPFeatureExtractor

feature_extractor = CLIPFeatureExtractor.from_pretrained(pipe_id, subfolder="feature_extractor")
```

Now, we need to implement the custom UNet. It's already available here: https://github.com/showlab/Show-1/blob/main/showone/models/unet_3d_condition.py. So, we create a Python script called `showone_unet_3d_condition.py` and copy over the implementation, changing the `UNet3DConditionModel` classname to `ShowOneUNet3DConditionModel` to avoid any conflicts with `diffusers`. This is because `diffusers` already has one [`UNet3DConditionModel`](https://github.com/huggingface/diffusers/blob/bc7a4d4917456afd70913be85bd25c556c25862c/src/diffusers/models/unet_3d_condition.py#L62). We put all the components needed to implement the class in `showone_unet_3d_condition.py` only. You can find the entire file here (TODO).

Once this is done, we can initialize the UNet:

```python
from showone_unet_3d_condition import ShowOneUNet3DConditionModel

unet = ShowOneUNet3DConditionModel.from_pretrained(pipe_id, subfolder="unet")
```

And then we implement the custom `TextToVideoIFPipeline` in another Python script: `pipeline_t2v_base_pixel.py`. 

Now that we have all the components, we can fully initialize the  `TextToVideoIFPipeline`:

```
from pipeline_t2v_base_pixel import TextToVideoIFPipeline
import torch

pipeline = TextToVideoIFPipeline(
    unet=unet, 
    text_encoder=text_encoder, 
    tokenizer=tokenizer, 
    scheduler=scheduler, 
    feature_extractor=feature_extractor
)
pipeline = pipeline.to(device="cuda")
pipeline.torch_dtype = torch.float16
```

For sharing with others, we can push this pipeline to the Hub:

```python
pipeline.push_to_hub("custom-t2v-pipeline")
```

After the pipeline is successfully pushed, we need to perform a couple of changes:

1. In `model_index.json` file, we need to change the `_class_name` attribute. It should be like so: https://huggingface.co/sayakpaul/show-1-base-with-code/blob/main/model_index.json#L2. 
2. Upload `showone_unet_3d_condition.py` to the `unet` directory ([example](https://huggingface.co/sayakpaul/show-1-base-with-code/blob/main/unet/showone_unet_3d_condition.py)). 
3. Upload `pipeline_t2v_base_pixel.py` to the pipeline base directory ([example](https://huggingface.co/sayakpaul/show-1-base-with-code/blob/main/pipeline_t2v_base_pixel.py)). 

Then we're ready for inference:

```python
from diffusers import DiffusionPipeline
import torch

pipeline = DiffusionPipeline.from_pretrained(
    "sayakpaul/show-1-base-with-code", trust_remote_code=True, torch_dtype=torch.float16
).to("cuda")

prompt = "hello"

# Text embeds
prompt_embeds, negative_embeds = pipeline.encode_prompt(prompt)

# Keyframes generation (8x64x40, 2fps)
video_frames = pipeline(
    prompt_embeds=prompt_embeds,
    negative_prompt_embeds=negative_embeds,
    num_frames=8,
    height=40,
    width=64,
    num_inference_steps=2,
    guidance_scale=9.0,
    output_type="pt"
).frames
```

File: src/diffusers/configuration_utils.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/configuration_utils.py:499]: to support a custom class

File: src/diffusers/pipelines/pipeline_utils.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/pipeline_utils.py:341]: We need to keep `hub_revision` as the other `revision` argument is already used for Git revisions

File: tests/pipelines/test_pipelines.py (hunk 3 above)
Comment (patrickvonplaten) [inline tests/pipelines/test_pipelines.py:886]:
```suggestion
            ""hf-internal-testing/tiny-sdxl-custom-components", custom_pipeline="my_pipeline"
```

File: tests/pipelines/test_pipelines.py [hunk 7]
```diff
@@ -862,6 +862,58 @@ def test_run_custom_pipeline(self):
...
+
+        pipeline = pipeline.to(torch_device)
+        images = pipeline("test", num_inference_steps=2, output_type="np")[0]
+
+        assert images.shape == (1, 64, 64, 3)
+
+        # Check that only loading custom componets "my_unet", "my_scheduler" and explicit custom pipeline works
+        pipeline = DiffusionPipeline.from_pretrained(
+            ""hf-internal-testing/tiny-sdxl-custom-components", custom_pipeline="my_pipeline"
```
Comment (patrickvonplaten) [inline tests/pipelines/test_pipelines.py:886]:
```suggestion
            "hf-internal-testing/tiny-sdxl-custom-components", custom_pipeline="my_pipeline"
```

**Comment by patrickvonplaten** [comment]
> Have been playing with this PR to support a bit more complicated pipelines such as: https://huggingface.co/showlab/show-1-base. Took a while to understand how the pipeline repository should be structured. So, documenting everything here.
> 
> If your pipeline has custom components that `diffusers` don't support already, you need to accompany the Python modules that implement them. These custom components could be VAE, UNet, scheduler, etc. For the text encoder, we rely on `transformers` anyway. So, that should be handled separately (more info [here](https://huggingface.co/docs/transformers/custom_models)). The pipeline code itself can be custom as well.
> 
> In case of `"showlab/show-1-base"`, we have a custom UNet and a custom pipeline (`TextToVideoIFPipeline`). For convenience, let's call the UNet `ShowOneUNet3DConditionModel`.
> 
> `"showlab/show-1-base"` already provides the checkpoints in the `diffusers` format, which is great. So, let's start loading up the components which are already well-supported:
> 
> 1. Text encoder:
> 
> ```python
> from transformers import T5Tokenizer, T5EncoderModel
> 
> pipe_id = "showlab/show-1-base"
> tokenizer = T5Tokenizer.from_pretrained(pipe_id, subfolder="tokenizer")
> text_encoder = T5EncoderModel.from_pretrained(pipe_id, subfolder="text_encoder")
> ```
> 
> 2. Scheduler:
> 
> ```python
> from diffusers import DPMSolverMultistepScheduler
> 
> scheduler = DPMSolverMultistepScheduler.from_pretrained(pipe_id, subfolder="scheduler")
> ```
> 
> 3. Image feature extractor:
> 
> ```python
> from transformers import CLIPFeatureExtractor
> 
> feature_extractor = CLIPFeatureExtractor.from_pretrained(pipe_id, subfolder="feature_extractor")
> ```
> 
> Now, we need to implement the custom UNet. It's already available here: https://github.com/showlab/Show-1/blob/main/showone/models/unet_3d_condition.py. So, we create a Python script called `showone_unet_3d_condition.py` and copy over the implementation, changing the `UNet3DConditionModel` classname to `ShowOneUNet3DConditionModel` to avoid any conflicts with `diffusers`. This is because `diffusers` already has one [`UNet3DConditionModel`](https://github.com/huggingface/diffusers/blob/bc7a4d4917456afd70913be85bd25c556c25862c/src/diffusers/models/unet_3d_condition.py#L62). We put all the components needed to implement the class in `showone_unet_3d_condition.py` only. You can find the entire file [here](https://huggingface.co/sayakpaul/show-1-base-with-code/blob/main/unet/showone_unet_3d_condition.py).
> 
> Once this is done, we can initialize the UNet:
> 
> ```python
> from showone_unet_3d_condition import ShowOneUNet3DConditionModel
> 
> unet = ShowOneUNet3DConditionModel.from_pretrained(pipe_id, subfolder="unet")
> ```
> 
> And then we implement the custom `TextToVideoIFPipeline` in another Python script: `pipeline_t2v_base_pixel.py`.
> 
> Now that we have all the components, we can fully initialize the `TextToVideoIFPipeline`:
> 
> ```
> from pipeline_t2v_base_pixel import TextToVideoIFPipeline
> import torch
> 
> pipeline = TextToVideoIFPipeline(
>     unet=unet, 
>     text_encoder=text_encoder, 
>     tokenizer=tokenizer, 
>     scheduler=scheduler, 
>     feature_extractor=feature_extractor
> )
> pipeline = pipeline.to(device="cuda")
> pipeline.torch_dtype = torch.float16
> ```
> 
> For sharing with others, we can push this pipeline to the Hub:
> 
> ```python
> pipeline.push_to_hub("custom-t2v-pipeline")
> ```
> 
> After the pipeline is successfully pushed, we need to perform a couple of changes:
> 
> 1. In `model_index.json` file, we need to change the `_class_name` attribute. It should be like so: https://huggingface.co/sayakpaul/show-1-base-with-code/blob/main/model_index.json#L2.
> 2. Upload `showone_unet_3d_condition.py` to the `unet` directory ([example](https://huggingface.co/sayakpaul/show-1-base-with-code/blob/main/unet/showone_unet_3d_condition.py)).
> 3. Upload `pipeline_t2v_base_pixel.py` to the pipeline base directory ([example](https://huggingface.co/sayakpaul/show-1-base-with-code/blob/main/pipeline_t2v_base_pixel.py)).
> 
> Then we're ready for inference:
> 
> ```python
> from diffusers import DiffusionPipeline
> import torch
> 
> pipeline = DiffusionPipeline.from_pretrained(
>     "sayakpaul/show-1-base-with-code", trust_remote_code=True, torch_dtype=torch.float16
> ).to("cuda")
> 
> prompt = "hello"
> 
> # Text embeds
> prompt_embeds, negative_embeds = pipeline.encode_prompt(prompt)
> 
> # Keyframes generation (8x64x40, 2fps)
> video_frames = pipeline(
>     prompt_embeds=prompt_embeds,
>     negative_prompt_embeds=negative_embeds,
>     num_frames=8,
>     height=40,
>     width=64,
>     num_inference_steps=2,
>     guidance_scale=9.0,
>     output_type="pt"
> ).frames
> ```

That's a great summary! Maybe we could make this a doc page in a follow-up PR? :-)

**Comment by sayakpaul** [comment]
@patrickvonplaten if you could look into https://github.com/huggingface/diffusers/pull/5491 before merging. Happy to create a doc after that.

**Comment by patrickvonplaten** [comment]
@sayakpaul Think it's ready for a final review :-)

**Review by sayakpaul — APPROVED** [review]
Okay for me to merge.

I will do a follow up after merge.
